File: dotnet/src/HybridRow/RowCursor.cs
// ------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All rights reserved.
// ------------------------------------------------------------

#pragma warning disable SA1307 // Accessible fields should begin with upper-case letter

// ReSharper disable InconsistentNaming
namespace Microsoft.Azure.Cosmos.Serialization.HybridRow
{
    using System.Diagnostics;
    using System.Runtime.CompilerServices;
    using Microsoft.Azure.Cosmos.Core;
    using Microsoft.Azure.Cosmos.Core.Utf8;
    using Microsoft.Azure.Cosmos.Serialization.HybridRow.Layouts;

    // ReSharper disable UseNameofExpression
    [DebuggerDisplay("{ToString()}")]
    public struct RowCursor
    {
        /// <summary>The layout describing the contents of the scope, or null if the scope is unschematized.</summary>
        internal Layout layout;

        /// <summary>The kind of scope within which this edit was prepared.</summary>
        internal LayoutScope scopeType;

        /// <summary>The type parameters of the scope within which this edit was prepared.</summary>
        internal TypeArgumentList scopeTypeArgs;

        /// <summary>If true, this scope's nested fields cannot be updated individually.</summary>
        /// <remarks>The entire scope can still be replaced.</remarks>
        internal bool immutable;

        /// <summary>If true, this scope is an unique index scope whose index will be built after its items are written.</remarks>
        internal bool deferUniqueIndex;

        /// <summary>
        /// The 0-based byte offset from the beginning of the row where the first sparse field within
        /// the scope begins.
        /// </summary>
        internal int start;

        /// <summary>True if an existing field matching the search criteria was found.</summary>
        internal bool exists;

        /// <summary>If existing, the scope relative path for writing.</summary>
        internal UtfAnyString writePath;

        /// <summary>If WritePath is tokenized, then its token.</summary>
        internal StringToken writePathToken;

        /// <summary>If existing, the offset scope relative path for reading.</summary>
        internal int pathOffset;

        /// <summary>If existing, the layout string token of scope relative path for reading.</summary>
        internal int pathToken;

        /// <summary>
        /// If existing, the offset to the metadata of the existing field, otherwise the location to
        /// insert a new field.
        /// </summary>
        internal int metaOffset;

        /// <summary>If existing, the layout code of the existing field, otherwise undefined.</summary>
        internal LayoutType cellType;

        /// <summary>If existing, the offset to the value of the existing field, otherwise undefined.</summary>
        internal int valueOffset;

        /// <summary>
        /// If existing, the offset to the end of the existing field. Used as a hint when skipping
        /// forward.
        /// </summary>
        internal int endOffset;

        /// <summary>For sized scopes (e.g. Typed Array), the number of elements.</summary>
        internal int count;

        /// <summary>For indexed scopes (e.g. Array), the 0-based index into the scope of the sparse field.</summary>
        internal int index;

        /// <summary>For types with generic parameters (e.g. <see cref="LayoutTuple" />, the type parameters.</summary>
        internal TypeArgumentList cellTypeArgs;

        public static RowCursor Create(ref RowBuffer row)
        {
            SchemaId schemaId = row.ReadSchemaId(1);
            Layout layout = row.Resolver.Resolve(schemaId);
            int sparseSegmentOffset = row.ComputeVariableValueOffset(layout, HybridRowHeader.Size, layout.NumVariable);
            return new RowCursor
            {
                layout = layout,
                scopeType = LayoutType.UDT,
                scopeTypeArgs = new TypeArgumentList(schemaId),
                start = HybridRowHeader.Size,
                metaOffset = sparseSegmentOffset,
                valueOffset = sparseSegmentOffset,
            };
        }

        public static ref RowCursor Create(ref RowBuffer row, out RowCursor cursor)
        {
            SchemaId schemaId = row.ReadSchemaId(1);
            Layout layout = row.Resolver.Resolve(schemaId);
            int sparseSegmentOffset = row.ComputeVariableValueOffset(layout, HybridRowHeader.Size, layout.NumVariable);
            cursor = new RowCursor
            {
                layout = layout,
                scopeType = LayoutType.UDT,
                scopeTypeArgs = new TypeArgumentList(schemaId),
                start = HybridRowHeader.Size,
                metaOffset = sparseSegmentOffset,
                valueOffset = sparseSegmentOffset,
            };

            return ref cursor;
        }

        public static ref RowCursor CreateForAppend(ref RowBuffer row, out RowCursor cursor)
        {
            SchemaId schemaId = row.ReadSchemaId(1);
            Layout layout = row.Resolver.Resolve(schemaId);
            cursor = new RowCursor
            {
                layout = layout,
                scopeType = LayoutType.UDT,
                scopeTypeArgs = new TypeArgumentList(schemaId),
                start = HybridRowHeader.Size,
                metaOffset = row.Length,
                valueOffset = row.Length,
            };

            return ref cursor;
        }

        /// <summary>For indexed scopes (e.g. Array), the 0-based index into the scope of the next insertion.</summary>
        [DebuggerBrowsable(DebuggerBrowsableState.Never)]
        public int Index
        {
            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            get => this.index;
        }

        /// <summary>If true, this scope's nested fields cannot be updated individually.</summary>
        /// <remarks>The entire scope can still be replaced.</remarks>
        [DebuggerBrowsable(DebuggerBrowsableState.Never)]
        public bool Immutable
        {
            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            get => this.immutable;
        }

        /// <summary>The kind of scope.</summary>
        [DebuggerBrowsable(DebuggerBrowsableState.Never)]
        public LayoutType ScopeType
        {
            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            get => this.scopeType;
        }

        /// <summary>For types with generic parameters (e.g. <see cref="LayoutTuple" />, the type parameters.</summary>
        [DebuggerBrowsable(DebuggerBrowsableState.Never)]
        public TypeArgumentList ScopeTypeArgs
        {
            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            get => this.scopeTypeArgs;
        }

        /// <summary>The layout describing the contents of the scope, or null if the scope is unschematized.</summary>
        [DebuggerBrowsable(DebuggerBrowsableState.Never)]
        public Layout Layout
        {
            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            get => this.layout;
        }

        /// <summary>The full logical type.</summary>
        [DebuggerBrowsable(DebuggerBrowsableState.Never)]
        public TypeArgument TypeArg
        {
            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            get => new TypeArgument(this.cellType, this.cellTypeArgs);
        }

        public override string ToString()
        {
            try
            {
                if (this.scopeType == null)
                {
                    return "<Invalid>";
                }

                TypeArgument scopeTypeArg = (this.scopeType == null) || (this.scopeType is LayoutEndScope)
                    ? default
                    : new TypeArgument(this.scopeType, this.scopeTypeArgs);

                TypeArgument typeArg = (this.cellType == null) || (this.cellType is LayoutEndScope)
                    ? default
                    : new TypeArgument(this.cellType, this.cellTypeArgs);

                string pathOrIndex = !this.writePath.IsNull ? this.writePath.ToString() : this.index.ToString();
                return $"{scopeTypeArg}[{pathOrIndex}] : {typeArg}@{this.metaOffset}/{this.valueOffset}" +
                       (this.immutable ? " immutable" : string.Empty);
            }
            catch
            {
                return "<???>";
            }
        }
    }

    public static class RowCursorExtensions
    {
        /// <summary>Makes a copy of the current cursor.</summary>
        /// <remarks>
        /// The two cursors will have independent and unconnected lifetimes after cloning.  However,
        /// mutations to a <see cref="RowBuffer" /> can invalidate any active cursors over the same row.
        /// </remarks>
        public static ref RowCursor Clone(this in RowCursor src, out RowCursor dest)
        {
            dest = src;
            return ref dest;
        }

        /// <summary>Returns an equivalent scope that is read-only.</summary>
        public static ref RowCursor AsReadOnly(this in RowCursor src, out RowCursor dest)
        {
            dest = src;
            dest.immutable = true;
            return ref dest;
        }

        public static ref RowCursor Find(this ref RowCursor edit, ref RowBuffer row, UtfAnyString path)
        {
            Contract.Requires(!edit.scopeType.IsIndexedScope);

            if (!(edit.cellType is LayoutEndScope))
            {
                while (row.SparseIteratorMoveNext(ref edit))
                {
                    if (path.Equals(row.ReadSparsePath(ref edit)))
                    {
                        edit.exists = true;
                        break;
                    }
                }
            }

            edit.writePath = path;
            edit.writePathToken = default;
            return ref edit;
        }

        public static ref RowCursor Find(this ref RowCursor edit, ref RowBuffer row, in StringToken pathToken)
        {
            Contract.Requires(!edit.scopeType.IsIndexedScope);

            if (!(edit.cellType is LayoutEndScope))
            {
                while (row.SparseIteratorMoveNext(ref edit))
                {
                    if (pathToken.Id == (ulong)edit.pathToken)
                    {
                        edit.exists = true;
                        break;
                    }
                }
            }

            edit.writePath = pathToken.Path;
            edit.writePathToken = pathToken;
            return ref edit;
        }

        public static bool MoveNext(this ref RowCursor edit, ref RowBuffer row)
        {
            edit.writePath = default;
            edit.writePathToken = default;
            return row.SparseIteratorMoveNext(ref edit);
        }

        public static bool MoveTo(this ref RowCursor edit, ref RowBuffer row, int index)
        {
            Contract.Assert(edit.index <= index);
            edit.writePath = default;
            edit.writePathToken = default;
            while (edit.index < index)
            {
                if (!row.SparseIteratorMoveNext(ref edit))
                {
                    return false;
                }
            }

            return true;
        }

        public static bool MoveNext(this ref RowCursor edit, ref RowBuffer row, ref RowCursor childScope)
        {
            if (childScope.scopeType != null)
            {
                edit.Skip(ref row, ref childScope);
            }

            return edit.MoveNext(ref row);
        }

        public static void Skip(this ref RowCursor edit, ref RowBuffer row, ref RowCursor childScope)
        {
            Contract.Requires(childScope.start == edit.valueOffset);
            if (!(childScope.cellType is LayoutEndScope))
            {
                while (row.SparseIteratorMoveNext(ref childScope))
                {
                }
            }

            if (childScope.scopeType.IsSizedScope)
            {
                edit.endOffset = childScope.metaOffset;
            }
            else
            {
                edit.endOffset = childScope.metaOffset + sizeof(LayoutCode); // Move past the end of scope marker.
            }

#if DEBUG
            childScope = default;
#endif
        }
    }
}

File: jre/src/main/java/com/azure/data/cosmos/serialization/hybridrow/RowCursor.java
//------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All rights reserved.
//------------------------------------------------------------

package com.azure.data.cosmos.serialization.hybridrow;

import com.azure.data.cosmos.core.Reference;
import com.azure.data.cosmos.core.UtfAnyString;
import com.azure.data.cosmos.serialization.hybridrow.layouts.Layout;
import com.azure.data.cosmos.serialization.hybridrow.layouts.LayoutEndScope;
import com.azure.data.cosmos.serialization.hybridrow.layouts.LayoutScope;
import com.azure.data.cosmos.serialization.hybridrow.layouts.LayoutTuple;
import com.azure.data.cosmos.serialization.hybridrow.layouts.LayoutType;
import com.azure.data.cosmos.serialization.hybridrow.layouts.StringToken;
import com.azure.data.cosmos.serialization.hybridrow.layouts.TypeArgument;
import com.azure.data.cosmos.serialization.hybridrow.layouts.TypeArgumentList;

import static com.google.common.base.Strings.lenientFormat;

public final class RowCursor {

    /**
     * If existing, the layout code of the existing field, otherwise undefined.
     */
    public LayoutType cellType;
    /**
     * For types with generic parameters (e.g. {@link LayoutTuple}, the type parameters.
     */
    public TypeArgumentList cellTypeArgs = new TypeArgumentList();
    /**
     * If true, this scope is an unique index scope whose index will be built after its items are written.
     */
    public boolean deferUniqueIndex;
    /**
     * If existing, the offset to the end of the existing field. Used as a hint when skipping
     * forward.
     */
    public int endOffset;
    /**
     * True if an existing field matching the search criteria was found.
     */
    public boolean exists;
    /**
     * If existing, the offset scope relative path for reading.
     */
    public int pathOffset;
    /**
     * If existing, the layout string token of scope relative path for reading.
     */
    public int pathToken;
    private int count;
    private boolean immutable;
    private int index;
    private Layout layout;
    private int metaOffset;
    private LayoutScope scopeType;
    private TypeArgumentList scopeTypeArgs = new TypeArgumentList();
    private int start;
    private int valueOffset;
    private UtfAnyString writePath;
    private StringToken writePathToken = new StringToken();

    public static RowCursor Create(Reference<RowBuffer> row) {

        final SchemaId schemaId = row.get().ReadSchemaId(1);
        final Layout layout = row.get().resolver().Resolve(schemaId);
        final int sparseSegmentOffset = row.get().computeVariableValueOffset(layout, HybridRowHeader.SIZE,
            layout.numVariable());

        final RowCursor cursor = new RowCursor()
            .layout(layout)
            .scopeType(LayoutType.UDT)
            .scopeTypeArgs(new TypeArgumentList(schemaId))
            .start(HybridRowHeader.SIZE)
            .metaOffset(sparseSegmentOffset)
            .valueOffset(sparseSegmentOffset);

        return cursor;
    }

    public static RowCursor Create(RowBuffer row) {

        final SchemaId schemaId = row.ReadSchemaId(1);
        final Layout layout = row.Resolver.Resolve(schemaId);
        final int sparseSegmentOffset = row.computeVariableValueOffset(layout, HybridRowHeader.Size,
            layout.NumVariable);

        return new RowCursor()
            .layout(layout)
            .scopeType(LayoutType.UDT)
            .scopeTypeArgs(new TypeArgumentList(schemaId, HybridRowHeader.SIZE, sparseSegmentOffset, sparseSegmentOffset);
    }

    /**
     * For sized scopes (e.g. Typed Array), the number of elements.
     */
    public int count() {
        return count;
    }

    public RowCursor count(int count) {
        this.count = count;
        return this;
    }

    // TODO: C# TO JAVA CONVERTER: 'ref return' methods are not converted by C# to Java Converter:
    //	public static ref RowCursor CreateForAppend(ref RowBuffer row, out RowCursor cursor)
    //		{
    //			SchemaId schemaId = row.ReadSchemaId(1);
    //			Layout layout = row.Resolver.Resolve(schemaId);
    //			cursor = new RowCursor { layout = layout, scopeType = LayoutType.UDT, scopeTypeArgs = new
    //			TypeArgumentList(schemaId), start = HybridRowHeader.Size, metaOffset = row.Length, valueOffset = row
    //			.Length};
    //
    //			return ref cursor;
    //		}

    /**
     * If {@code true}, this scope's nested fields cannot be updated individually
     * <p>
     * The entire scope can still be replaced.
     */
    public boolean immutable() {
        return immutable;
    }

    public RowCursor immutable(boolean immutable) {
        this.immutable = immutable;
        return this;
    }

    /**
     * For indexed scopes (e.g. an Array scope), the zero-based index into the scope of the sparse field
     */
    public int index() {
        return index;
    }

    public RowCursor index(int index) {
        this.index = index;
        return this;
    }

    /**
     * The layout describing the contents of the scope, or {@code null} if the scope is unschematized.
     */
    public Layout layout() {
        return layout;
    }

    public RowCursor layout(Layout layout) {
        this.layout = layout;
        return this;
    }

    /**
     * If existing, the offset to the metadata of the existing field, otherwise the location to
     * insert a new field.
     */
    public int metaOffset() {
        return metaOffset;
    }

    public RowCursor metaOffset(int metaOffset) {
        this.metaOffset = metaOffset;
        return this;
    }

    /**
     * The kind of scope within which this edit was prepared
     */
    public LayoutScope scopeType() {
        return scopeType;
    }

    public RowCursor scopeType(LayoutScope scopeType) {
        this.scopeType = scopeType;
        return this;
    }

    /**
     * The type parameters of the scope within which this edit was prepared
     */
    public TypeArgumentList scopeTypeArgs() {
        return scopeTypeArgs;
    }

    public RowCursor scopeTypeArgs(TypeArgumentList scopeTypeArgs) {
        this.scopeTypeArgs = scopeTypeArgs;
        return this;
    }

    /**
     * The 0-based byte offset from the beginning of the row where the first sparse field within
     * the scope begins.
     */
    public int start() {
        return start;
    }

    public RowCursor start(int start) {
        this.start = start;
        return this;
    }

    @Override
    public String toString() {

        try {

            if (this.scopeType() == null) {
                return "<Invalid>";
            }

            TypeArgument scopeTypeArg = (this.scopeType() instanceof LayoutEndScope)
                ? new TypeArgument()
                : new TypeArgument(this.scopeType(), this.scopeTypeArgs().clone());

            TypeArgument typeArg = (this.cellType == null) || (this.cellType instanceof LayoutEndScope)
                ? new TypeArgument()
                : new TypeArgument(this.cellType, this.cellTypeArgs.clone());

            String pathOrIndex = this.writePath().isNull() ? String.valueOf(this.index()) : this.writePath().toString();

            return lenientFormat("%s[%s] : %s@%s/%s%s",
                scopeTypeArg.clone(),
                pathOrIndex,
                typeArg.clone(),
                this.metaOffset(),
                this.valueOffset(),
                this.immutable() ? " immutable" : "");

        } catch (Exception ignored) {
            return "<???>";
        }
    }

    /**
     * If existing, the offset to the value of the existing field, otherwise undefined.
     */
    public int valueOffset() {
        return valueOffset;
    }

    public RowCursor valueOffset(int valueOffset) {
        this.valueOffset = valueOffset;
        return this;
    }

    /**
     * If existing, the scope relative path for writing.
     */
    public UtfAnyString writePath() {
        return writePath;
    }

    public void writePath(UtfAnyString writePath) {
        this.writePath = writePath;
    }

    /**
     * If WritePath is tokenized, then its token.
     */
    public StringToken writePathToken() {
        return writePathToken;
    }

    public void writePathToken(StringToken writePathToken) {
        this.writePathToken = writePathToken;
    }
}